dlib/media/ffmpeg_utils.cpp

```diff
@@ -0,0 +1,763 @@
+                                                            0, NULL));
+                    int ret = 0;
+                    if ((ret = swr_init(audioResamplerCtx_.get())) < 0)
+                    {
+                        audioResamplerCtx_.reset(nullptr);
+                        throw std::runtime_error("swr_init() failed : " + get_av_error(ret));
+                    }
+                }
+            }
+        }
+
+        void sw_audio_resampler::resize(
+            const Frame&    src,
+            int             dst_sample_rate,
+            uint64_t        dst_channel_layout,
+            AVSampleFormat  dst_samplefmt,
+            Frame&          dst
+        )
+        {
+            using namespace std::chrono;
+            assert(src.is_audio());
+
+            const bool is_same_object = std::addressof(src) == std::addressof(dst);
+
+            reset(src.frame->sample_rate, src.frame->channel_layout, (AVSampleFormat)src.frame->format,
+                dst_sample_rate, dst_channel_layout, dst_samplefmt);
+
+            if (audioResamplerCtx_)
+            {
+                Frame tmp;
+                Frame* ptr = std::addressof(dst);
+
+                const int64_t delay       = swr_get_delay(audioResamplerCtx_.get(), src_sample_rate_);
+                const auto dst_nb_samples = av_rescale_rnd(delay + src.frame->nb_samples, dst_sample_rate_, src_sample_rate_, AV_ROUND_UP);
+
+                if (is_same_object ||
+                    !dst.is_audio() ||
+                    std::tie(dst.frame->sample_rate, dst.frame->channel_layout, dst.frame->format, dst.frame->nb_samples) !=
+                    std::tie(dst_sample_rate_, dst_channel_layout_, dst_fmt_, dst_nb_samples))
+                {
+                    tmp = Frame::make_audio(dst_sample_rate_, dst_nb_samples, dst_channel_layout_, dst_fmt_, std::chrono::system_clock::time_point{});
+                    ptr = std::addressof(tmp);
+                }
+
+                int ret = swr_convert(audioResamplerCtx_.get(),
+                                    ptr->frame->data, ptr->frame->nb_samples,
+                                    (const uint8_t**)src.frame->data, src.frame->nb_samples);
+                if (ret < 0)
+                    throw std::runtime_error("swr_convert() failed : " + get_av_error(ret));
+
+                ptr->frame->nb_samples = ret;
+                ptr->frame->pts   = tracked_samples_;
+                ptr->timestamp    = system_clock::time_point{nanoseconds{av_rescale_q(tracked_samples_,
+                                                                                    {1, dst_sample_rate_},
+                                                                                    {nanoseconds::period::num, nanoseconds::period::den})}};
+                tracked_samples_ += ptr->frame->nb_samples;
+
+                if (ptr!= std::addressof(dst))
+                    dst = std::move(tmp);
```

Comment:
Maybe just replace with:
```
av_ptr<AVFrame> tmp = make_avframe();
                tmp->sample_rate    = dst_sample_rate;
                tmp->channel_layout = dst_channel_layout;
                tmp->format         = (int)dst_fmt;

                const int ret = swr_convert_frame(audioResamplerCtx.get(), tmp.get(), src.frame.get());
                if (ret < 0)
                    throw std::runtime_error("swr_convert_frame() failed : " + get_av_error(ret));

                dst.frame       = std::move(tmp);
                dst.frame->pts  = tracked_samples;
                dst.timestamp   = system_clock::time_point{nanoseconds{av_rescale_q(tracked_samples,
                                                                                    {1, dst_sample_rate},
                                                                                    {nanoseconds::period::num, nanoseconds::period::den})}};
                tracked_samples += dst.frame->nb_samples;
```

Reply:
it's a slight pessimization in the case where `dst` already has the right buffer sizes allocated, but the code is a lot cleaner and easier to reason about.